src/commands.rs

```diff
@@ -0,0 +1,1173 @@
+    let mut args = args.into_iter().skip(1);
+    let key = args.next_string()?;
+
+    let mut paths: Vec<Path> = vec![];
+    let mut format = Format::JSON;
+    let mut indent = String::new();
+    let mut space = String::new();
+    let mut newline = String::new();
+    while let Ok(arg) = args.next_string() {
+        match arg {
+            // fast way to consider arg a path by using the max length of all possible subcommands
+            // See #390 for the comparison of this function with/without this optimization
+            arg if arg.len() > JSONGET_SUBCOMMANDS_MAXSTRLEN => paths.push(Path::new(arg)),
+            arg if arg.eq_ignore_ascii_case(CMD_ARG_INDENT) => indent = args.next_string()?,
+            arg if arg.eq_ignore_ascii_case(CMD_ARG_NEWLINE) => newline = args.next_string()?,
+            arg if arg.eq_ignore_ascii_case(CMD_ARG_SPACE) => space = args.next_string()?,
+            // Silently ignore. Compatibility with ReJSON v1.0 which has this option. See #168 TODO add support
+            arg if arg.eq_ignore_ascii_case(CMD_ARG_NOESCAPE) => continue,
+            arg if arg.eq_ignore_ascii_case(CMD_ARG_FORMAT) => {
+                format = Format::from_str(args.next_string()?.as_str())?
+            }
+            _ => paths.push(Path::new(arg)),
+        };
+    }
+
+    // path is optional -> no path found we use root "$"
+    if paths.is_empty() {
+        paths.push(Path::new(JSON_ROOT_PATH.to_string()));
+    }
+
+    let key = manager.open_key_read(ctx, &key)?;
+    let value = match key.get_value()? {
+        Some(doc) => KeyValue::new(doc)
+            .to_json(&mut paths, indent, newline, space, format)?
+            .into(),
+        None => RedisValue::Null,
+    };
+
+    Ok(value)
+}
+
+pub fn command_json_set<M: Manager>(manager: M, ctx: &Context, args: Vec<String>) -> RedisResult {
+    let mut args = args.into_iter().skip(1);
+
+    let key = args.next_string()?;
+    let path = backwards_compat_path(args.next_string()?);
+    let value = args.next_string()?;
+
+    let mut format = Format::JSON;
+    let mut set_option = SetOptions::None;
+
+    while let Some(s) = args.next() {
+        match s.to_uppercase().as_str() {
+            "NX" => set_option = SetOptions::NotExists,
+            "XX" => set_option = SetOptions::AlreadyExists,
+            "FORMAT" => {
+                format = Format::from_str(args.next_string()?.as_str())?;
+            }
+            _ => break,
+        };
+    }
+
+    let mut redis_key = manager.open_key_write(ctx, &key)?;
+    let current = redis_key.get_value()?;
+
+    let val = manager.from_str(&value, format)?;
+
+    match (current, set_option) {
+        (Some(ref mut doc), ref op) => {
+            if path == JSON_ROOT_PATH {
+                if *op != SetOptions::NotExists {
+                    redis_key.set_value(Vec::new(), val)?;
+                    redis_key.apply_changes(ctx, "json.set")?;
```

Comment:
Should `apply_changes` be called only when the return value from `redis_key.set_value` is `Ok(true)`?
(although currently it seems to be possible to get only `Ok(false)` or `Err`)

Reply:
Well, its updating the root so it can not failed (return false). We can assert it, WDYT?

Reply:
There are more calls to `apply_changes`, some consider the result of the "write" operation (set_value, del, etc.),
But some don't, or only consider the last result of a `for` iteration, such as in `command_json_num_op` , `command_json_bool_toggle` , `command_json_str_append` , `command_json_arr_append` , `command_json_arr_insert` , `command_json_arr_trim` , `command_json_clear` 
However, this is also related to atomicity and the behavior needs to be decided as commented [here](https://github.com/RedisJSON/RedisJSON/pull/336#discussion_r659807305)
(also for `command_json_arr_pop`)

Reply:
Yes, the assumption is that the last result should be the same as the first result. but I agree this all story need to be handled together with the API changed